{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 420,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.codonalign.codonseq import CodonSeq, cal_dn_ds\n",
    "import Bio.Data.CodonTable\n",
    "from Bio import Entrez,SeqIO\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We going to calculate for each codon how many positions can lead to synonymous mutations and how many to non-synonymous mutations. We will use the genetic code to determine the amino acid encoded by each codon. We will then calculate the number of codons that encode the same amino acid as the original codon and the number of codons that encode a different amino acid. Finally, we will calculate the ratio of synonymous to non-synonymous mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 421,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "# Import the standard codon table, those are the codons which are used to translate the DNA sequence into amino acids at human cells\n",
    "print(Bio.Data.CodonTable.standard_dna_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non-synonymous mutation: TTT -> ATT\n",
      "Non-synonymous mutation: TTT -> CTT\n",
      "Non-synonymous mutation: TTT -> GTT\n",
      "Non-synonymous mutation: TTT -> TAT\n",
      "Non-synonymous mutation: TTT -> TCT\n",
      "Non-synonymous mutation: TTT -> TGT\n",
      "Non-synonymous mutation: TTT -> TTA\n",
      "Synonymous mutation: TTT -> TTC\n",
      "Non-synonymous mutation: TTT -> TTG\n",
      "Non-synonymous mutation: TTC -> ATC\n",
      "Non-synonymous mutation: TTC -> CTC\n",
      "Non-synonymous mutation: TTC -> GTC\n",
      "Non-synonymous mutation: TTC -> TAC\n",
      "Non-synonymous mutation: TTC -> TCC\n",
      "Non-synonymous mutation: TTC -> TGC\n",
      "Non-synonymous mutation: TTC -> TTA\n",
      "Non-synonymous mutation: TTC -> TTG\n",
      "Synonymous mutation: TTC -> TTT\n",
      "Non-synonymous mutation: TTA -> ATA\n",
      "Synonymous mutation: TTA -> CTA\n",
      "Non-synonymous mutation: TTA -> GTA\n",
      " Non-synonymous mutation to Stop codon!: TTA -> TAA\n",
      "Non-synonymous mutation: TTA -> TCA\n",
      " Non-synonymous mutation to Stop codon!: TTA -> TGA\n",
      "Non-synonymous mutation: TTA -> TTC\n",
      "Synonymous mutation: TTA -> TTG\n",
      "Non-synonymous mutation: TTA -> TTT\n",
      "Non-synonymous mutation: TTG -> ATG\n",
      "Synonymous mutation: TTG -> CTG\n",
      "Non-synonymous mutation: TTG -> GTG\n",
      " Non-synonymous mutation to Stop codon!: TTG -> TAG\n",
      "Non-synonymous mutation: TTG -> TCG\n",
      "Non-synonymous mutation: TTG -> TGG\n",
      "Synonymous mutation: TTG -> TTA\n",
      "Non-synonymous mutation: TTG -> TTC\n",
      "Non-synonymous mutation: TTG -> TTT\n",
      "Non-synonymous mutation: TCT -> ACT\n",
      "Non-synonymous mutation: TCT -> CCT\n",
      "Non-synonymous mutation: TCT -> GCT\n",
      "Non-synonymous mutation: TCT -> TAT\n",
      "Non-synonymous mutation: TCT -> TGT\n",
      "Non-synonymous mutation: TCT -> TTT\n",
      "Synonymous mutation: TCT -> TCA\n",
      "Synonymous mutation: TCT -> TCC\n",
      "Synonymous mutation: TCT -> TCG\n",
      "Non-synonymous mutation: TCC -> ACC\n",
      "Non-synonymous mutation: TCC -> CCC\n",
      "Non-synonymous mutation: TCC -> GCC\n",
      "Non-synonymous mutation: TCC -> TAC\n",
      "Non-synonymous mutation: TCC -> TGC\n",
      "Non-synonymous mutation: TCC -> TTC\n",
      "Synonymous mutation: TCC -> TCA\n",
      "Synonymous mutation: TCC -> TCG\n",
      "Synonymous mutation: TCC -> TCT\n",
      "Non-synonymous mutation: TCA -> ACA\n",
      "Non-synonymous mutation: TCA -> CCA\n",
      "Non-synonymous mutation: TCA -> GCA\n",
      " Non-synonymous mutation to Stop codon!: TCA -> TAA\n",
      " Non-synonymous mutation to Stop codon!: TCA -> TGA\n",
      "Non-synonymous mutation: TCA -> TTA\n",
      "Synonymous mutation: TCA -> TCC\n",
      "Synonymous mutation: TCA -> TCG\n",
      "Synonymous mutation: TCA -> TCT\n",
      "Non-synonymous mutation: TCG -> ACG\n",
      "Non-synonymous mutation: TCG -> CCG\n",
      "Non-synonymous mutation: TCG -> GCG\n",
      " Non-synonymous mutation to Stop codon!: TCG -> TAG\n",
      "Non-synonymous mutation: TCG -> TGG\n",
      "Non-synonymous mutation: TCG -> TTG\n",
      "Synonymous mutation: TCG -> TCA\n",
      "Synonymous mutation: TCG -> TCC\n",
      "Synonymous mutation: TCG -> TCT\n",
      "Non-synonymous mutation: TAT -> AAT\n",
      "Non-synonymous mutation: TAT -> CAT\n",
      "Non-synonymous mutation: TAT -> GAT\n",
      "Non-synonymous mutation: TAT -> TCT\n",
      "Non-synonymous mutation: TAT -> TGT\n",
      "Non-synonymous mutation: TAT -> TTT\n",
      " Non-synonymous mutation to Stop codon!: TAT -> TAA\n",
      "Synonymous mutation: TAT -> TAC\n",
      " Non-synonymous mutation to Stop codon!: TAT -> TAG\n",
      "Non-synonymous mutation: TAC -> AAC\n",
      "Non-synonymous mutation: TAC -> CAC\n",
      "Non-synonymous mutation: TAC -> GAC\n",
      "Non-synonymous mutation: TAC -> TCC\n",
      "Non-synonymous mutation: TAC -> TGC\n",
      "Non-synonymous mutation: TAC -> TTC\n",
      " Non-synonymous mutation to Stop codon!: TAC -> TAA\n",
      " Non-synonymous mutation to Stop codon!: TAC -> TAG\n",
      "Synonymous mutation: TAC -> TAT\n",
      "Non-synonymous mutation: TGT -> AGT\n",
      "Non-synonymous mutation: TGT -> CGT\n",
      "Non-synonymous mutation: TGT -> GGT\n",
      "Non-synonymous mutation: TGT -> TAT\n",
      "Non-synonymous mutation: TGT -> TCT\n",
      "Non-synonymous mutation: TGT -> TTT\n",
      " Non-synonymous mutation to Stop codon!: TGT -> TGA\n",
      "Synonymous mutation: TGT -> TGC\n",
      "Non-synonymous mutation: TGT -> TGG\n",
      "Non-synonymous mutation: TGC -> AGC\n",
      "Non-synonymous mutation: TGC -> CGC\n",
      "Non-synonymous mutation: TGC -> GGC\n",
      "Non-synonymous mutation: TGC -> TAC\n",
      "Non-synonymous mutation: TGC -> TCC\n",
      "Non-synonymous mutation: TGC -> TTC\n",
      " Non-synonymous mutation to Stop codon!: TGC -> TGA\n",
      "Non-synonymous mutation: TGC -> TGG\n",
      "Synonymous mutation: TGC -> TGT\n",
      "Non-synonymous mutation: TGG -> AGG\n",
      "Non-synonymous mutation: TGG -> CGG\n",
      "Non-synonymous mutation: TGG -> GGG\n",
      " Non-synonymous mutation to Stop codon!: TGG -> TAG\n",
      "Non-synonymous mutation: TGG -> TCG\n",
      "Non-synonymous mutation: TGG -> TTG\n",
      " Non-synonymous mutation to Stop codon!: TGG -> TGA\n",
      "Non-synonymous mutation: TGG -> TGC\n",
      "Non-synonymous mutation: TGG -> TGT\n",
      "Non-synonymous mutation: CTT -> ATT\n",
      "Non-synonymous mutation: CTT -> GTT\n",
      "Non-synonymous mutation: CTT -> TTT\n",
      "Non-synonymous mutation: CTT -> CAT\n",
      "Non-synonymous mutation: CTT -> CCT\n",
      "Non-synonymous mutation: CTT -> CGT\n",
      "Synonymous mutation: CTT -> CTA\n",
      "Synonymous mutation: CTT -> CTC\n",
      "Synonymous mutation: CTT -> CTG\n",
      "Non-synonymous mutation: CTC -> ATC\n",
      "Non-synonymous mutation: CTC -> GTC\n",
      "Non-synonymous mutation: CTC -> TTC\n",
      "Non-synonymous mutation: CTC -> CAC\n",
      "Non-synonymous mutation: CTC -> CCC\n",
      "Non-synonymous mutation: CTC -> CGC\n",
      "Synonymous mutation: CTC -> CTA\n",
      "Synonymous mutation: CTC -> CTG\n",
      "Synonymous mutation: CTC -> CTT\n",
      "Non-synonymous mutation: CTA -> ATA\n",
      "Non-synonymous mutation: CTA -> GTA\n",
      "Synonymous mutation: CTA -> TTA\n",
      "Non-synonymous mutation: CTA -> CAA\n",
      "Non-synonymous mutation: CTA -> CCA\n",
      "Non-synonymous mutation: CTA -> CGA\n",
      "Synonymous mutation: CTA -> CTC\n",
      "Synonymous mutation: CTA -> CTG\n",
      "Synonymous mutation: CTA -> CTT\n",
      "Non-synonymous mutation: CTG -> ATG\n",
      "Non-synonymous mutation: CTG -> GTG\n",
      "Synonymous mutation: CTG -> TTG\n",
      "Non-synonymous mutation: CTG -> CAG\n",
      "Non-synonymous mutation: CTG -> CCG\n",
      "Non-synonymous mutation: CTG -> CGG\n",
      "Synonymous mutation: CTG -> CTA\n",
      "Synonymous mutation: CTG -> CTC\n",
      "Synonymous mutation: CTG -> CTT\n",
      "Non-synonymous mutation: CCT -> ACT\n",
      "Non-synonymous mutation: CCT -> GCT\n",
      "Non-synonymous mutation: CCT -> TCT\n",
      "Non-synonymous mutation: CCT -> CAT\n",
      "Non-synonymous mutation: CCT -> CGT\n",
      "Non-synonymous mutation: CCT -> CTT\n",
      "Synonymous mutation: CCT -> CCA\n",
      "Synonymous mutation: CCT -> CCC\n",
      "Synonymous mutation: CCT -> CCG\n",
      "Non-synonymous mutation: CCC -> ACC\n",
      "Non-synonymous mutation: CCC -> GCC\n",
      "Non-synonymous mutation: CCC -> TCC\n",
      "Non-synonymous mutation: CCC -> CAC\n",
      "Non-synonymous mutation: CCC -> CGC\n",
      "Non-synonymous mutation: CCC -> CTC\n",
      "Synonymous mutation: CCC -> CCA\n",
      "Synonymous mutation: CCC -> CCG\n",
      "Synonymous mutation: CCC -> CCT\n",
      "Non-synonymous mutation: CCA -> ACA\n",
      "Non-synonymous mutation: CCA -> GCA\n",
      "Non-synonymous mutation: CCA -> TCA\n",
      "Non-synonymous mutation: CCA -> CAA\n",
      "Non-synonymous mutation: CCA -> CGA\n",
      "Non-synonymous mutation: CCA -> CTA\n",
      "Synonymous mutation: CCA -> CCC\n",
      "Synonymous mutation: CCA -> CCG\n",
      "Synonymous mutation: CCA -> CCT\n",
      "Non-synonymous mutation: CCG -> ACG\n",
      "Non-synonymous mutation: CCG -> GCG\n",
      "Non-synonymous mutation: CCG -> TCG\n",
      "Non-synonymous mutation: CCG -> CAG\n",
      "Non-synonymous mutation: CCG -> CGG\n",
      "Non-synonymous mutation: CCG -> CTG\n",
      "Synonymous mutation: CCG -> CCA\n",
      "Synonymous mutation: CCG -> CCC\n",
      "Synonymous mutation: CCG -> CCT\n",
      "Non-synonymous mutation: CAT -> AAT\n",
      "Non-synonymous mutation: CAT -> GAT\n",
      "Non-synonymous mutation: CAT -> TAT\n",
      "Non-synonymous mutation: CAT -> CCT\n",
      "Non-synonymous mutation: CAT -> CGT\n",
      "Non-synonymous mutation: CAT -> CTT\n",
      "Non-synonymous mutation: CAT -> CAA\n",
      "Synonymous mutation: CAT -> CAC\n",
      "Non-synonymous mutation: CAT -> CAG\n",
      "Non-synonymous mutation: CAC -> AAC\n",
      "Non-synonymous mutation: CAC -> GAC\n",
      "Non-synonymous mutation: CAC -> TAC\n",
      "Non-synonymous mutation: CAC -> CCC\n",
      "Non-synonymous mutation: CAC -> CGC\n",
      "Non-synonymous mutation: CAC -> CTC\n",
      "Non-synonymous mutation: CAC -> CAA\n",
      "Non-synonymous mutation: CAC -> CAG\n",
      "Synonymous mutation: CAC -> CAT\n",
      "Non-synonymous mutation: CAA -> AAA\n",
      "Non-synonymous mutation: CAA -> GAA\n",
      " Non-synonymous mutation to Stop codon!: CAA -> TAA\n",
      "Non-synonymous mutation: CAA -> CCA\n",
      "Non-synonymous mutation: CAA -> CGA\n",
      "Non-synonymous mutation: CAA -> CTA\n",
      "Non-synonymous mutation: CAA -> CAC\n",
      "Synonymous mutation: CAA -> CAG\n",
      "Non-synonymous mutation: CAA -> CAT\n",
      "Non-synonymous mutation: CAG -> AAG\n",
      "Non-synonymous mutation: CAG -> GAG\n",
      " Non-synonymous mutation to Stop codon!: CAG -> TAG\n",
      "Non-synonymous mutation: CAG -> CCG\n",
      "Non-synonymous mutation: CAG -> CGG\n",
      "Non-synonymous mutation: CAG -> CTG\n",
      "Synonymous mutation: CAG -> CAA\n",
      "Non-synonymous mutation: CAG -> CAC\n",
      "Non-synonymous mutation: CAG -> CAT\n",
      "Non-synonymous mutation: CGT -> AGT\n",
      "Non-synonymous mutation: CGT -> GGT\n",
      "Non-synonymous mutation: CGT -> TGT\n",
      "Non-synonymous mutation: CGT -> CAT\n",
      "Non-synonymous mutation: CGT -> CCT\n",
      "Non-synonymous mutation: CGT -> CTT\n",
      "Synonymous mutation: CGT -> CGA\n",
      "Synonymous mutation: CGT -> CGC\n",
      "Synonymous mutation: CGT -> CGG\n",
      "Non-synonymous mutation: CGC -> AGC\n",
      "Non-synonymous mutation: CGC -> GGC\n",
      "Non-synonymous mutation: CGC -> TGC\n",
      "Non-synonymous mutation: CGC -> CAC\n",
      "Non-synonymous mutation: CGC -> CCC\n",
      "Non-synonymous mutation: CGC -> CTC\n",
      "Synonymous mutation: CGC -> CGA\n",
      "Synonymous mutation: CGC -> CGG\n",
      "Synonymous mutation: CGC -> CGT\n",
      "Synonymous mutation: CGA -> AGA\n",
      "Non-synonymous mutation: CGA -> GGA\n",
      " Non-synonymous mutation to Stop codon!: CGA -> TGA\n",
      "Non-synonymous mutation: CGA -> CAA\n",
      "Non-synonymous mutation: CGA -> CCA\n",
      "Non-synonymous mutation: CGA -> CTA\n",
      "Synonymous mutation: CGA -> CGC\n",
      "Synonymous mutation: CGA -> CGG\n",
      "Synonymous mutation: CGA -> CGT\n",
      "Synonymous mutation: CGG -> AGG\n",
      "Non-synonymous mutation: CGG -> GGG\n",
      "Non-synonymous mutation: CGG -> TGG\n",
      "Non-synonymous mutation: CGG -> CAG\n",
      "Non-synonymous mutation: CGG -> CCG\n",
      "Non-synonymous mutation: CGG -> CTG\n",
      "Synonymous mutation: CGG -> CGA\n",
      "Synonymous mutation: CGG -> CGC\n",
      "Synonymous mutation: CGG -> CGT\n",
      "Non-synonymous mutation: ATT -> CTT\n",
      "Non-synonymous mutation: ATT -> GTT\n",
      "Non-synonymous mutation: ATT -> TTT\n",
      "Non-synonymous mutation: ATT -> AAT\n",
      "Non-synonymous mutation: ATT -> ACT\n",
      "Non-synonymous mutation: ATT -> AGT\n",
      "Synonymous mutation: ATT -> ATA\n",
      "Synonymous mutation: ATT -> ATC\n",
      "Non-synonymous mutation: ATT -> ATG\n",
      "Non-synonymous mutation: ATC -> CTC\n",
      "Non-synonymous mutation: ATC -> GTC\n",
      "Non-synonymous mutation: ATC -> TTC\n",
      "Non-synonymous mutation: ATC -> AAC\n",
      "Non-synonymous mutation: ATC -> ACC\n",
      "Non-synonymous mutation: ATC -> AGC\n",
      "Synonymous mutation: ATC -> ATA\n",
      "Non-synonymous mutation: ATC -> ATG\n",
      "Synonymous mutation: ATC -> ATT\n",
      "Non-synonymous mutation: ATA -> CTA\n",
      "Non-synonymous mutation: ATA -> GTA\n",
      "Non-synonymous mutation: ATA -> TTA\n",
      "Non-synonymous mutation: ATA -> AAA\n",
      "Non-synonymous mutation: ATA -> ACA\n",
      "Non-synonymous mutation: ATA -> AGA\n",
      "Synonymous mutation: ATA -> ATC\n",
      "Non-synonymous mutation: ATA -> ATG\n",
      "Synonymous mutation: ATA -> ATT\n",
      "Non-synonymous mutation: ATG -> CTG\n",
      "Non-synonymous mutation: ATG -> GTG\n",
      "Non-synonymous mutation: ATG -> TTG\n",
      "Non-synonymous mutation: ATG -> AAG\n",
      "Non-synonymous mutation: ATG -> ACG\n",
      "Non-synonymous mutation: ATG -> AGG\n",
      "Non-synonymous mutation: ATG -> ATA\n",
      "Non-synonymous mutation: ATG -> ATC\n",
      "Non-synonymous mutation: ATG -> ATT\n",
      "Non-synonymous mutation: ACT -> CCT\n",
      "Non-synonymous mutation: ACT -> GCT\n",
      "Non-synonymous mutation: ACT -> TCT\n",
      "Non-synonymous mutation: ACT -> AAT\n",
      "Non-synonymous mutation: ACT -> AGT\n",
      "Non-synonymous mutation: ACT -> ATT\n",
      "Synonymous mutation: ACT -> ACA\n",
      "Synonymous mutation: ACT -> ACC\n",
      "Synonymous mutation: ACT -> ACG\n",
      "Non-synonymous mutation: ACC -> CCC\n",
      "Non-synonymous mutation: ACC -> GCC\n",
      "Non-synonymous mutation: ACC -> TCC\n",
      "Non-synonymous mutation: ACC -> AAC\n",
      "Non-synonymous mutation: ACC -> AGC\n",
      "Non-synonymous mutation: ACC -> ATC\n",
      "Synonymous mutation: ACC -> ACA\n",
      "Synonymous mutation: ACC -> ACG\n",
      "Synonymous mutation: ACC -> ACT\n",
      "Non-synonymous mutation: ACA -> CCA\n",
      "Non-synonymous mutation: ACA -> GCA\n",
      "Non-synonymous mutation: ACA -> TCA\n",
      "Non-synonymous mutation: ACA -> AAA\n",
      "Non-synonymous mutation: ACA -> AGA\n",
      "Non-synonymous mutation: ACA -> ATA\n",
      "Synonymous mutation: ACA -> ACC\n",
      "Synonymous mutation: ACA -> ACG\n",
      "Synonymous mutation: ACA -> ACT\n",
      "Non-synonymous mutation: ACG -> CCG\n",
      "Non-synonymous mutation: ACG -> GCG\n",
      "Non-synonymous mutation: ACG -> TCG\n",
      "Non-synonymous mutation: ACG -> AAG\n",
      "Non-synonymous mutation: ACG -> AGG\n",
      "Non-synonymous mutation: ACG -> ATG\n",
      "Synonymous mutation: ACG -> ACA\n",
      "Synonymous mutation: ACG -> ACC\n",
      "Synonymous mutation: ACG -> ACT\n",
      "Non-synonymous mutation: AAT -> CAT\n",
      "Non-synonymous mutation: AAT -> GAT\n",
      "Non-synonymous mutation: AAT -> TAT\n",
      "Non-synonymous mutation: AAT -> ACT\n",
      "Non-synonymous mutation: AAT -> AGT\n",
      "Non-synonymous mutation: AAT -> ATT\n",
      "Non-synonymous mutation: AAT -> AAA\n",
      "Synonymous mutation: AAT -> AAC\n",
      "Non-synonymous mutation: AAT -> AAG\n",
      "Non-synonymous mutation: AAC -> CAC\n",
      "Non-synonymous mutation: AAC -> GAC\n",
      "Non-synonymous mutation: AAC -> TAC\n",
      "Non-synonymous mutation: AAC -> ACC\n",
      "Non-synonymous mutation: AAC -> AGC\n",
      "Non-synonymous mutation: AAC -> ATC\n",
      "Non-synonymous mutation: AAC -> AAA\n",
      "Non-synonymous mutation: AAC -> AAG\n",
      "Synonymous mutation: AAC -> AAT\n",
      "Non-synonymous mutation: AAA -> CAA\n",
      "Non-synonymous mutation: AAA -> GAA\n",
      " Non-synonymous mutation to Stop codon!: AAA -> TAA\n",
      "Non-synonymous mutation: AAA -> ACA\n",
      "Non-synonymous mutation: AAA -> AGA\n",
      "Non-synonymous mutation: AAA -> ATA\n",
      "Non-synonymous mutation: AAA -> AAC\n",
      "Synonymous mutation: AAA -> AAG\n",
      "Non-synonymous mutation: AAA -> AAT\n",
      "Non-synonymous mutation: AAG -> CAG\n",
      "Non-synonymous mutation: AAG -> GAG\n",
      " Non-synonymous mutation to Stop codon!: AAG -> TAG\n",
      "Non-synonymous mutation: AAG -> ACG\n",
      "Non-synonymous mutation: AAG -> AGG\n",
      "Non-synonymous mutation: AAG -> ATG\n",
      "Synonymous mutation: AAG -> AAA\n",
      "Non-synonymous mutation: AAG -> AAC\n",
      "Non-synonymous mutation: AAG -> AAT\n",
      "Non-synonymous mutation: AGT -> CGT\n",
      "Non-synonymous mutation: AGT -> GGT\n",
      "Non-synonymous mutation: AGT -> TGT\n",
      "Non-synonymous mutation: AGT -> AAT\n",
      "Non-synonymous mutation: AGT -> ACT\n",
      "Non-synonymous mutation: AGT -> ATT\n",
      "Non-synonymous mutation: AGT -> AGA\n",
      "Synonymous mutation: AGT -> AGC\n",
      "Non-synonymous mutation: AGT -> AGG\n",
      "Non-synonymous mutation: AGC -> CGC\n",
      "Non-synonymous mutation: AGC -> GGC\n",
      "Non-synonymous mutation: AGC -> TGC\n",
      "Non-synonymous mutation: AGC -> AAC\n",
      "Non-synonymous mutation: AGC -> ACC\n",
      "Non-synonymous mutation: AGC -> ATC\n",
      "Non-synonymous mutation: AGC -> AGA\n",
      "Non-synonymous mutation: AGC -> AGG\n",
      "Synonymous mutation: AGC -> AGT\n",
      "Synonymous mutation: AGA -> CGA\n",
      "Non-synonymous mutation: AGA -> GGA\n",
      " Non-synonymous mutation to Stop codon!: AGA -> TGA\n",
      "Non-synonymous mutation: AGA -> AAA\n",
      "Non-synonymous mutation: AGA -> ACA\n",
      "Non-synonymous mutation: AGA -> ATA\n",
      "Non-synonymous mutation: AGA -> AGC\n",
      "Synonymous mutation: AGA -> AGG\n",
      "Non-synonymous mutation: AGA -> AGT\n",
      "Synonymous mutation: AGG -> CGG\n",
      "Non-synonymous mutation: AGG -> GGG\n",
      "Non-synonymous mutation: AGG -> TGG\n",
      "Non-synonymous mutation: AGG -> AAG\n",
      "Non-synonymous mutation: AGG -> ACG\n",
      "Non-synonymous mutation: AGG -> ATG\n",
      "Synonymous mutation: AGG -> AGA\n",
      "Non-synonymous mutation: AGG -> AGC\n",
      "Non-synonymous mutation: AGG -> AGT\n",
      "Non-synonymous mutation: GTT -> ATT\n",
      "Non-synonymous mutation: GTT -> CTT\n",
      "Non-synonymous mutation: GTT -> TTT\n",
      "Non-synonymous mutation: GTT -> GAT\n",
      "Non-synonymous mutation: GTT -> GCT\n",
      "Non-synonymous mutation: GTT -> GGT\n",
      "Synonymous mutation: GTT -> GTA\n",
      "Synonymous mutation: GTT -> GTC\n",
      "Synonymous mutation: GTT -> GTG\n",
      "Non-synonymous mutation: GTC -> ATC\n",
      "Non-synonymous mutation: GTC -> CTC\n",
      "Non-synonymous mutation: GTC -> TTC\n",
      "Non-synonymous mutation: GTC -> GAC\n",
      "Non-synonymous mutation: GTC -> GCC\n",
      "Non-synonymous mutation: GTC -> GGC\n",
      "Synonymous mutation: GTC -> GTA\n",
      "Synonymous mutation: GTC -> GTG\n",
      "Synonymous mutation: GTC -> GTT\n",
      "Non-synonymous mutation: GTA -> ATA\n",
      "Non-synonymous mutation: GTA -> CTA\n",
      "Non-synonymous mutation: GTA -> TTA\n",
      "Non-synonymous mutation: GTA -> GAA\n",
      "Non-synonymous mutation: GTA -> GCA\n",
      "Non-synonymous mutation: GTA -> GGA\n",
      "Synonymous mutation: GTA -> GTC\n",
      "Synonymous mutation: GTA -> GTG\n",
      "Synonymous mutation: GTA -> GTT\n",
      "Non-synonymous mutation: GTG -> ATG\n",
      "Non-synonymous mutation: GTG -> CTG\n",
      "Non-synonymous mutation: GTG -> TTG\n",
      "Non-synonymous mutation: GTG -> GAG\n",
      "Non-synonymous mutation: GTG -> GCG\n",
      "Non-synonymous mutation: GTG -> GGG\n",
      "Synonymous mutation: GTG -> GTA\n",
      "Synonymous mutation: GTG -> GTC\n",
      "Synonymous mutation: GTG -> GTT\n",
      "Non-synonymous mutation: GCT -> ACT\n",
      "Non-synonymous mutation: GCT -> CCT\n",
      "Non-synonymous mutation: GCT -> TCT\n",
      "Non-synonymous mutation: GCT -> GAT\n",
      "Non-synonymous mutation: GCT -> GGT\n",
      "Non-synonymous mutation: GCT -> GTT\n",
      "Synonymous mutation: GCT -> GCA\n",
      "Synonymous mutation: GCT -> GCC\n",
      "Synonymous mutation: GCT -> GCG\n",
      "Non-synonymous mutation: GCC -> ACC\n",
      "Non-synonymous mutation: GCC -> CCC\n",
      "Non-synonymous mutation: GCC -> TCC\n",
      "Non-synonymous mutation: GCC -> GAC\n",
      "Non-synonymous mutation: GCC -> GGC\n",
      "Non-synonymous mutation: GCC -> GTC\n",
      "Synonymous mutation: GCC -> GCA\n",
      "Synonymous mutation: GCC -> GCG\n",
      "Synonymous mutation: GCC -> GCT\n",
      "Non-synonymous mutation: GCA -> ACA\n",
      "Non-synonymous mutation: GCA -> CCA\n",
      "Non-synonymous mutation: GCA -> TCA\n",
      "Non-synonymous mutation: GCA -> GAA\n",
      "Non-synonymous mutation: GCA -> GGA\n",
      "Non-synonymous mutation: GCA -> GTA\n",
      "Synonymous mutation: GCA -> GCC\n",
      "Synonymous mutation: GCA -> GCG\n",
      "Synonymous mutation: GCA -> GCT\n",
      "Non-synonymous mutation: GCG -> ACG\n",
      "Non-synonymous mutation: GCG -> CCG\n",
      "Non-synonymous mutation: GCG -> TCG\n",
      "Non-synonymous mutation: GCG -> GAG\n",
      "Non-synonymous mutation: GCG -> GGG\n",
      "Non-synonymous mutation: GCG -> GTG\n",
      "Synonymous mutation: GCG -> GCA\n",
      "Synonymous mutation: GCG -> GCC\n",
      "Synonymous mutation: GCG -> GCT\n",
      "Non-synonymous mutation: GAT -> AAT\n",
      "Non-synonymous mutation: GAT -> CAT\n",
      "Non-synonymous mutation: GAT -> TAT\n",
      "Non-synonymous mutation: GAT -> GCT\n",
      "Non-synonymous mutation: GAT -> GGT\n",
      "Non-synonymous mutation: GAT -> GTT\n",
      "Non-synonymous mutation: GAT -> GAA\n",
      "Synonymous mutation: GAT -> GAC\n",
      "Non-synonymous mutation: GAT -> GAG\n",
      "Non-synonymous mutation: GAC -> AAC\n",
      "Non-synonymous mutation: GAC -> CAC\n",
      "Non-synonymous mutation: GAC -> TAC\n",
      "Non-synonymous mutation: GAC -> GCC\n",
      "Non-synonymous mutation: GAC -> GGC\n",
      "Non-synonymous mutation: GAC -> GTC\n",
      "Non-synonymous mutation: GAC -> GAA\n",
      "Non-synonymous mutation: GAC -> GAG\n",
      "Synonymous mutation: GAC -> GAT\n",
      "Non-synonymous mutation: GAA -> AAA\n",
      "Non-synonymous mutation: GAA -> CAA\n",
      " Non-synonymous mutation to Stop codon!: GAA -> TAA\n",
      "Non-synonymous mutation: GAA -> GCA\n",
      "Non-synonymous mutation: GAA -> GGA\n",
      "Non-synonymous mutation: GAA -> GTA\n",
      "Non-synonymous mutation: GAA -> GAC\n",
      "Synonymous mutation: GAA -> GAG\n",
      "Non-synonymous mutation: GAA -> GAT\n",
      "Non-synonymous mutation: GAG -> AAG\n",
      "Non-synonymous mutation: GAG -> CAG\n",
      " Non-synonymous mutation to Stop codon!: GAG -> TAG\n",
      "Non-synonymous mutation: GAG -> GCG\n",
      "Non-synonymous mutation: GAG -> GGG\n",
      "Non-synonymous mutation: GAG -> GTG\n",
      "Synonymous mutation: GAG -> GAA\n",
      "Non-synonymous mutation: GAG -> GAC\n",
      "Non-synonymous mutation: GAG -> GAT\n",
      "Non-synonymous mutation: GGT -> AGT\n",
      "Non-synonymous mutation: GGT -> CGT\n",
      "Non-synonymous mutation: GGT -> TGT\n",
      "Non-synonymous mutation: GGT -> GAT\n",
      "Non-synonymous mutation: GGT -> GCT\n",
      "Non-synonymous mutation: GGT -> GTT\n",
      "Synonymous mutation: GGT -> GGA\n",
      "Synonymous mutation: GGT -> GGC\n",
      "Synonymous mutation: GGT -> GGG\n",
      "Non-synonymous mutation: GGC -> AGC\n",
      "Non-synonymous mutation: GGC -> CGC\n",
      "Non-synonymous mutation: GGC -> TGC\n",
      "Non-synonymous mutation: GGC -> GAC\n",
      "Non-synonymous mutation: GGC -> GCC\n",
      "Non-synonymous mutation: GGC -> GTC\n",
      "Synonymous mutation: GGC -> GGA\n",
      "Synonymous mutation: GGC -> GGG\n",
      "Synonymous mutation: GGC -> GGT\n",
      "Non-synonymous mutation: GGA -> AGA\n",
      "Non-synonymous mutation: GGA -> CGA\n",
      " Non-synonymous mutation to Stop codon!: GGA -> TGA\n",
      "Non-synonymous mutation: GGA -> GAA\n",
      "Non-synonymous mutation: GGA -> GCA\n",
      "Non-synonymous mutation: GGA -> GTA\n",
      "Synonymous mutation: GGA -> GGC\n",
      "Synonymous mutation: GGA -> GGG\n",
      "Synonymous mutation: GGA -> GGT\n",
      "Non-synonymous mutation: GGG -> AGG\n",
      "Non-synonymous mutation: GGG -> CGG\n",
      "Non-synonymous mutation: GGG -> TGG\n",
      "Non-synonymous mutation: GGG -> GAG\n",
      "Non-synonymous mutation: GGG -> GCG\n",
      "Non-synonymous mutation: GGG -> GTG\n",
      "Synonymous mutation: GGG -> GGA\n",
      "Synonymous mutation: GGG -> GGC\n",
      "Synonymous mutation: GGG -> GGT\n",
      "Stop codon Non-synonymous mutation: TAA  -> AAA\n",
      "Stop codon Non-synonymous mutation: TAA  -> CAA\n",
      "Stop codon Non-synonymous mutation: TAA  -> GAA\n",
      "Stop codon Non-synonymous mutation: TAA  -> TCA\n",
      "Stop codon Synonymous mutation: TAA  -> TGA\n",
      "Stop codon Non-synonymous mutation: TAA  -> TTA\n",
      "Stop codon Non-synonymous mutation: TAA  -> TAC\n",
      "Stop codon Synonymous mutation: TAA  -> TAG\n",
      "Stop codon Non-synonymous mutation: TAA  -> TAT\n",
      "Stop codon Non-synonymous mutation: TAG  -> AAG\n",
      "Stop codon Non-synonymous mutation: TAG  -> CAG\n",
      "Stop codon Non-synonymous mutation: TAG  -> GAG\n",
      "Stop codon Non-synonymous mutation: TAG  -> TCG\n",
      "Stop codon Non-synonymous mutation: TAG  -> TGG\n",
      "Stop codon Non-synonymous mutation: TAG  -> TTG\n",
      "Stop codon Synonymous mutation: TAG  -> TAA\n",
      "Stop codon Non-synonymous mutation: TAG  -> TAC\n",
      "Stop codon Non-synonymous mutation: TAG  -> TAT\n",
      "Stop codon Non-synonymous mutation: TGA  -> AGA\n",
      "Stop codon Non-synonymous mutation: TGA  -> CGA\n",
      "Stop codon Non-synonymous mutation: TGA  -> GGA\n",
      "Stop codon Synonymous mutation: TGA  -> TAA\n",
      "Stop codon Non-synonymous mutation: TGA  -> TCA\n",
      "Stop codon Non-synonymous mutation: TGA  -> TTA\n",
      "Stop codon Non-synonymous mutation: TGA  -> TGC\n",
      "Stop codon Non-synonymous mutation: TGA  -> TGG\n",
      "Stop codon Non-synonymous mutation: TGA  -> TGT\n"
     ]
    }
   ],
   "source": [
    "# For each codon we going to calculate how many positions can lead to synonymous mutations and how many to non-synonymous mutations\n",
    "# We going to examen all possible mutations for each codon and calculate the number of synonymous and non-synonymous mutations\n",
    "# For example, for the codon 'TTC' which codes for the amino acid 'F' (Phenylalanine) we have 9 possible mutations\n",
    "# ATC, CTC, GTC, TAC, TCC, TGC, TTA, TTG which lead to non-synonymous mutations and TTT which leads to a synonymous mutation\n",
    "\n",
    "# Initialize a dictionary to store the number of synonymous positions and for each codon\n",
    "synonymous_positions = {}\n",
    "\n",
    "for codon in Bio.Data.CodonTable.standard_dna_table.forward_table:\n",
    "    nucleotides = ['A', 'C', 'G', 'T']\n",
    "    for i in range(3):\n",
    "        for nucleotide in nucleotides:\n",
    "            if codon[i] != nucleotide:\n",
    "                mutated_codon = codon[:i] + nucleotide + codon[i+1:]\n",
    "                if mutated_codon in Bio.Data.CodonTable.standard_dna_table.stop_codons: #Non-stop codons mutation to a stop codon are non-synonymous\n",
    "                    print(f\" Non-synonymous mutation to Stop codon!: {codon} -> {mutated_codon}\") \n",
    "                elif Bio.Data.CodonTable.standard_dna_table.forward_table[codon] == Bio.Data.CodonTable.standard_dna_table.forward_table[mutated_codon]: \n",
    "                    print(f\"Synonymous mutation: {codon} -> {mutated_codon}\")\n",
    "                    synonymous_positions[codon] = synonymous_positions.get(codon, 0) + 1\n",
    "                else:\n",
    "                    print(f\"Non-synonymous mutation: {codon} -> {mutated_codon}\")\n",
    "\n",
    "for codon in Bio.Data.CodonTable.standard_dna_table.stop_codons:\n",
    "    nucleotides = ['A', 'C', 'G', 'T']\n",
    "    for i in range(3):\n",
    "        for nucleotide in nucleotides:\n",
    "            if codon[i] != nucleotide:\n",
    "                mutated_codon = codon[:i] + nucleotide + codon[i+1:]\n",
    "                if mutated_codon in Bio.Data.CodonTable.standard_dna_table.stop_codons:\n",
    "                    print(f\"Stop codon Synonymous mutation: {codon}  -> {mutated_codon}\")\n",
    "                    synonymous_positions[codon] = synonymous_positions.get(codon, 0) + 1\n",
    "                else:\n",
    "                    print(f\"Stop codon Non-synonymous mutation: {codon}  -> {mutated_codon}\")\n",
    "\n",
    "\n",
    "                \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 423,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': 1,\n",
       " 'TTC': 1,\n",
       " 'TTA': 2,\n",
       " 'TTG': 2,\n",
       " 'TCT': 3,\n",
       " 'TCC': 3,\n",
       " 'TCA': 3,\n",
       " 'TCG': 3,\n",
       " 'TAT': 1,\n",
       " 'TAC': 1,\n",
       " 'TGT': 1,\n",
       " 'TGC': 1,\n",
       " 'CTT': 3,\n",
       " 'CTC': 3,\n",
       " 'CTA': 4,\n",
       " 'CTG': 4,\n",
       " 'CCT': 3,\n",
       " 'CCC': 3,\n",
       " 'CCA': 3,\n",
       " 'CCG': 3,\n",
       " 'CAT': 1,\n",
       " 'CAC': 1,\n",
       " 'CAA': 1,\n",
       " 'CAG': 1,\n",
       " 'CGT': 3,\n",
       " 'CGC': 3,\n",
       " 'CGA': 4,\n",
       " 'CGG': 4,\n",
       " 'ATT': 2,\n",
       " 'ATC': 2,\n",
       " 'ATA': 2,\n",
       " 'ACT': 3,\n",
       " 'ACC': 3,\n",
       " 'ACA': 3,\n",
       " 'ACG': 3,\n",
       " 'AAT': 1,\n",
       " 'AAC': 1,\n",
       " 'AAA': 1,\n",
       " 'AAG': 1,\n",
       " 'AGT': 1,\n",
       " 'AGC': 1,\n",
       " 'AGA': 2,\n",
       " 'AGG': 2,\n",
       " 'GTT': 3,\n",
       " 'GTC': 3,\n",
       " 'GTA': 3,\n",
       " 'GTG': 3,\n",
       " 'GCT': 3,\n",
       " 'GCC': 3,\n",
       " 'GCA': 3,\n",
       " 'GCG': 3,\n",
       " 'GAT': 1,\n",
       " 'GAC': 1,\n",
       " 'GAA': 1,\n",
       " 'GAG': 1,\n",
       " 'GGT': 3,\n",
       " 'GGC': 3,\n",
       " 'GGA': 3,\n",
       " 'GGG': 3,\n",
       " 'TAA': 2,\n",
       " 'TAG': 1,\n",
       " 'TGA': 1}"
      ]
     },
     "execution_count": 423,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synonymous_positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 424,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 29-APR-2021\n",
      "\n",
      "Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 20-FEB-2024\n"
     ]
    }
   ],
   "source": [
    "# Importing a GeneBank file of the Covid-19 genome\n",
    "Entrez.email = \"jcecomputationalbiology@gmail.com\" \n",
    "\n",
    "# Importing the Covid-19 genome updated on 29-APR-2021\n",
    "with Entrez.efetch(db=\"nucleotide\", id=\"MZ054892.1\", rettype=\"gb\", retmode=\"text\") as handle:\n",
    "    older_genbank_record = SeqIO.read(handle, \"genbank\")\n",
    "    older_genbank_record_name = older_genbank_record.annotations[\"source\"] + \" \" + older_genbank_record.annotations[\"date\"]\n",
    "    print(older_genbank_record_name)\n",
    "\n",
    "print()\n",
    "# Importing the Covid-19 genome updated on 20-FEB-2024\n",
    "with Entrez.efetch(db=\"nucleotide\", id=\"PP348372.1\", rettype=\"gb\", retmode=\"text\") as handle:\n",
    "    newer_genbank_record = SeqIO.read(handle, \"genbank\")\n",
    "    newer_genbank_record_name = newer_genbank_record.annotations[\"source\"] + \" \" + newer_genbank_record.annotations[\"date\"]\n",
    "    print(newer_genbank_record_name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "א. אורך כל גנום"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 29-APR-2021 has 29740 nucleotides\n",
      "Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 20-FEB-2024 has 29661 nucleotides\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Calculating the length of each genome\n",
    "print(f\"{older_genbank_record_name} has {len(older_genbank_record.seq)} nucleotides\")\n",
    "print(f\"{newer_genbank_record_name} has {len(newer_genbank_record.seq)} nucleotides\")\n",
    "\n",
    "print(type(older_genbank_record.features[2].location) == Bio.SeqFeature.CompoundLocation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ב. כמה גנים יש בכל אחד מהם? מתוכם כמה גנים מקודדים לחלבונים?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elements in Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 29-APR-2021\n",
      "Number of source: 1\n",
      "Number of gene: 11\n",
      "Number of CDS: 12\n",
      "Number of mat_peptide: 26\n",
      "Number of stem_loop: 4\n",
      "\n",
      "Elements in Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) 20-FEB-2024\n",
      "Number of source: 1\n",
      "Number of gene: 11\n",
      "Number of CDS: 12\n",
      "Number of mat_peptide: 26\n",
      "Number of stem_loop: 5\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def calculate_num_of_gens_by_type(genbank_record):\n",
    "    elements_count = {}\n",
    "    for feature in genbank_record.features:\n",
    "        feature_type = feature.type\n",
    "        # if type(feature.location) == Bio.SeqFeature.CompoundLocation:\n",
    "        #     for sub_feature in feature.location.parts:\n",
    "        #         # elements_count[feature_type] = elements_count.get(feature_type, 0) + 1\n",
    "        #         continue\n",
    "        # else:\n",
    "        elements_count[feature_type] = elements_count.get(feature_type, 0) + 1\n",
    "\n",
    "    for element_type, count in elements_count.items():\n",
    "        print(f\"Number of {element_type}: {count}\")\n",
    "\n",
    "print(f\"Elements in {older_genbank_record_name}\")\n",
    "calculate_num_of_gens_by_type(older_genbank_record)\n",
    "print()\n",
    "print(f\"Elements in {newer_genbank_record_name}\")\n",
    "calculate_num_of_gens_by_type(newer_genbank_record)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ג. כמה גנים משותפים יש ביניהם"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Length</th>\n",
       "      <th>Product</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Translation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13194</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>234</td>\n",
       "      <td>13427</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>8088</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>13427</td>\n",
       "      <td>21514</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13209</td>\n",
       "      <td>ORF1a polyprotein</td>\n",
       "      <td>234</td>\n",
       "      <td>13442</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>3813</td>\n",
       "      <td>surface glycoprotein</td>\n",
       "      <td>21522</td>\n",
       "      <td>25334</td>\n",
       "      <td>1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYXXXXXXXXX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>828</td>\n",
       "      <td>ORF3a protein</td>\n",
       "      <td>25343</td>\n",
       "      <td>26170</td>\n",
       "      <td>1</td>\n",
       "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E</td>\n",
       "      <td>228</td>\n",
       "      <td>envelope protein</td>\n",
       "      <td>26195</td>\n",
       "      <td>26422</td>\n",
       "      <td>1</td>\n",
       "      <td>MYSFVSEETGTLIVNSVLLFFAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>M</td>\n",
       "      <td>669</td>\n",
       "      <td>membrane glycoprotein</td>\n",
       "      <td>26473</td>\n",
       "      <td>27141</td>\n",
       "      <td>1</td>\n",
       "      <td>MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>183</td>\n",
       "      <td>ORF6 protein</td>\n",
       "      <td>27152</td>\n",
       "      <td>27334</td>\n",
       "      <td>1</td>\n",
       "      <td>MHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF7a protein</td>\n",
       "      <td>27341</td>\n",
       "      <td>27706</td>\n",
       "      <td>1</td>\n",
       "      <td>MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ORF7b</td>\n",
       "      <td>132</td>\n",
       "      <td>ORF7b</td>\n",
       "      <td>27703</td>\n",
       "      <td>27834</td>\n",
       "      <td>1</td>\n",
       "      <td>MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF8 protein</td>\n",
       "      <td>27841</td>\n",
       "      <td>28206</td>\n",
       "      <td>1</td>\n",
       "      <td>MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>N</td>\n",
       "      <td>1257</td>\n",
       "      <td>nucleocapsid phosphoprotein</td>\n",
       "      <td>28221</td>\n",
       "      <td>29477</td>\n",
       "      <td>1</td>\n",
       "      <td>MYNGPQNQRNGPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLPN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ORF10</td>\n",
       "      <td>117</td>\n",
       "      <td>ORF10 protein</td>\n",
       "      <td>29502</td>\n",
       "      <td>29618</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene  Length                      Product  Start    End  Strand  \\\n",
       "0   ORF1ab   13194           ORF1ab polyprotein    234  13427       1   \n",
       "1   ORF1ab    8088           ORF1ab polyprotein  13427  21514       1   \n",
       "2   ORF1ab   13209            ORF1a polyprotein    234  13442       1   \n",
       "3        S    3813         surface glycoprotein  21522  25334       1   \n",
       "4    ORF3a     828                ORF3a protein  25343  26170       1   \n",
       "5        E     228             envelope protein  26195  26422       1   \n",
       "6        M     669        membrane glycoprotein  26473  27141       1   \n",
       "7     ORF6     183                 ORF6 protein  27152  27334       1   \n",
       "8    ORF7a     366                ORF7a protein  27341  27706       1   \n",
       "9    ORF7b     132                        ORF7b  27703  27834       1   \n",
       "10    ORF8     366                 ORF8 protein  27841  28206       1   \n",
       "11       N    1257  nucleocapsid phosphoprotein  28221  29477       1   \n",
       "12   ORF10     117                ORF10 protein  29502  29618       1   \n",
       "\n",
       "                                          Translation  \n",
       "0   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "1   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "2   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "3   MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYXXXXXXXXX...  \n",
       "4   MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...  \n",
       "5   MYSFVSEETGTLIVNSVLLFFAFVVFLLVTLAILTALRLCAYCCNI...  \n",
       "6   MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...  \n",
       "7   MHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTEN...  \n",
       "8   MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...  \n",
       "9         MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA  \n",
       "10  MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...  \n",
       "11  MYNGPQNQRNGPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLPN...  \n",
       "12             MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT  "
      ]
     },
     "execution_count": 427,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculating the number of genes in each genome and how many genes translated to proteins\n",
    "\n",
    "# For the older genome\n",
    "older_genome_df = pd.DataFrame(columns=[\"Gene\", \"Length\", \"Product\", \"Start\", \"End\", \"Strand\", \"Translation\"])\n",
    "newer_genome_df = pd.DataFrame(columns=[\"Gene\", \"Length\", \"Product\", \"Start\", \"End\", \"Strand\", \"Translation\"])\n",
    "\n",
    "# Creating a dataframe to store the results for each genome\n",
    "def from_genbank_to_df(genbank_record, genome_df):\n",
    "    for feature in genbank_record.features:\n",
    "        if feature.type == \"CDS\":\n",
    "            gene = feature.qualifiers.get(\"gene\", [\"\"])[0]\n",
    "            length = len(feature)\n",
    "            product = feature.qualifiers.get(\"product\", [\"\"])[0]\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            strand = feature.location.strand\n",
    "            translation = feature.qualifiers.get(\"translation\", [\"\"])[0]\n",
    "            #sequence = str(genbank_record.seq[start:end])\n",
    "            if type(feature.location) == Bio.SeqFeature.CompoundLocation:\n",
    "                for sub_feature in feature.location.parts:\n",
    "                    start = sub_feature.start\n",
    "                    end = sub_feature.end\n",
    "                    length = len(sub_feature)\n",
    "                    #sequence = str(genbank_record.seq[start:end])\n",
    "                    genome_df.loc[len(genome_df)] = [gene, length, product, start+1, end, strand, translation]\n",
    "            else:\n",
    "                genome_df.loc[len(genome_df)] = [gene, length, product, start+1, end, strand, translation]\n",
    "\n",
    "\n",
    "from_genbank_to_df(older_genbank_record, older_genome_df)\n",
    "from_genbank_to_df(newer_genbank_record, newer_genome_df)\n",
    "# remove doplicates\n",
    "older_genome_df.drop_duplicates(subset=[\"Start\", \"End\"], inplace=True)\n",
    "newer_genome_df.drop_duplicates(subset=[\"Start\", \"End\"], inplace=True)\n",
    "\n",
    "\n",
    "older_genome_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Length</th>\n",
       "      <th>Product</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Translation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13194</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>164</td>\n",
       "      <td>13357</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>8088</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>13357</td>\n",
       "      <td>21444</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13209</td>\n",
       "      <td>ORF1a polyprotein</td>\n",
       "      <td>164</td>\n",
       "      <td>13372</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>3810</td>\n",
       "      <td>surface glycoprotein</td>\n",
       "      <td>21452</td>\n",
       "      <td>25261</td>\n",
       "      <td>1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>828</td>\n",
       "      <td>ORF3a protein</td>\n",
       "      <td>25270</td>\n",
       "      <td>26097</td>\n",
       "      <td>1</td>\n",
       "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>E</td>\n",
       "      <td>228</td>\n",
       "      <td>envelope protein</td>\n",
       "      <td>26122</td>\n",
       "      <td>26349</td>\n",
       "      <td>1</td>\n",
       "      <td>MYSFVSEEIGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>M</td>\n",
       "      <td>669</td>\n",
       "      <td>membrane glycoprotein</td>\n",
       "      <td>26400</td>\n",
       "      <td>27068</td>\n",
       "      <td>1</td>\n",
       "      <td>MAHSNGTITVEELKKLLEEWNLVIGFLFLAWICLLQFAYANRNRFL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>186</td>\n",
       "      <td>ORF6 protein</td>\n",
       "      <td>27079</td>\n",
       "      <td>27264</td>\n",
       "      <td>1</td>\n",
       "      <td>MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF7a protein</td>\n",
       "      <td>27271</td>\n",
       "      <td>27636</td>\n",
       "      <td>1</td>\n",
       "      <td>MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ORF7b</td>\n",
       "      <td>132</td>\n",
       "      <td>ORF7b</td>\n",
       "      <td>27633</td>\n",
       "      <td>27764</td>\n",
       "      <td>1</td>\n",
       "      <td>MIELSLIDFYLCFLAFLLLLVLIMLIIFWFSLELQDHNETCHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF8 protein</td>\n",
       "      <td>27771</td>\n",
       "      <td>28136</td>\n",
       "      <td>1</td>\n",
       "      <td>MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>N</td>\n",
       "      <td>1251</td>\n",
       "      <td>nucleocapsid phosphoprotein</td>\n",
       "      <td>28151</td>\n",
       "      <td>29401</td>\n",
       "      <td>1</td>\n",
       "      <td>MSDNGPQNQRNALRITFGGPSDSTGSNQNGGARSKQRRPQGLPNNT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ORF10</td>\n",
       "      <td>117</td>\n",
       "      <td>ORF10 protein</td>\n",
       "      <td>29426</td>\n",
       "      <td>29542</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene  Length                      Product  Start    End  Strand  \\\n",
       "0   ORF1ab   13194           ORF1ab polyprotein    164  13357       1   \n",
       "1   ORF1ab    8088           ORF1ab polyprotein  13357  21444       1   \n",
       "2   ORF1ab   13209            ORF1a polyprotein    164  13372       1   \n",
       "3        S    3810         surface glycoprotein  21452  25261       1   \n",
       "4    ORF3a     828                ORF3a protein  25270  26097       1   \n",
       "5        E     228             envelope protein  26122  26349       1   \n",
       "6        M     669        membrane glycoprotein  26400  27068       1   \n",
       "7     ORF6     186                 ORF6 protein  27079  27264       1   \n",
       "8    ORF7a     366                ORF7a protein  27271  27636       1   \n",
       "9    ORF7b     132                        ORF7b  27633  27764       1   \n",
       "10    ORF8     366                 ORF8 protein  27771  28136       1   \n",
       "11       N    1251  nucleocapsid phosphoprotein  28151  29401       1   \n",
       "12   ORF10     117                ORF10 protein  29426  29542       1   \n",
       "\n",
       "                                          Translation  \n",
       "0   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "1   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "2   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "3   MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...  \n",
       "4   MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...  \n",
       "5   MYSFVSEEIGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...  \n",
       "6   MAHSNGTITVEELKKLLEEWNLVIGFLFLAWICLLQFAYANRNRFL...  \n",
       "7   MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...  \n",
       "8   MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...  \n",
       "9         MIELSLIDFYLCFLAFLLLLVLIMLIIFWFSLELQDHNETCHA  \n",
       "10  MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...  \n",
       "11  MSDNGPQNQRNALRITFGGPSDSTGSNQNGGARSKQRRPQGLPNNT...  \n",
       "12             MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT  "
      ]
     },
     "execution_count": 428,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "newer_genome_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 429,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Length_x</th>\n",
       "      <th>Product</th>\n",
       "      <th>Start_x</th>\n",
       "      <th>End_x</th>\n",
       "      <th>Strand_x</th>\n",
       "      <th>Translation_x</th>\n",
       "      <th>Length_y</th>\n",
       "      <th>Start_y</th>\n",
       "      <th>End_y</th>\n",
       "      <th>Strand_y</th>\n",
       "      <th>Translation_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13194</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>234</td>\n",
       "      <td>13427</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>13194</td>\n",
       "      <td>164</td>\n",
       "      <td>13357</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13194</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>234</td>\n",
       "      <td>13427</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>8088</td>\n",
       "      <td>13357</td>\n",
       "      <td>21444</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>8088</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>13427</td>\n",
       "      <td>21514</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>13194</td>\n",
       "      <td>164</td>\n",
       "      <td>13357</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>8088</td>\n",
       "      <td>ORF1ab polyprotein</td>\n",
       "      <td>13427</td>\n",
       "      <td>21514</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>8088</td>\n",
       "      <td>13357</td>\n",
       "      <td>21444</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>13209</td>\n",
       "      <td>ORF1a polyprotein</td>\n",
       "      <td>234</td>\n",
       "      <td>13442</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>13209</td>\n",
       "      <td>164</td>\n",
       "      <td>13372</td>\n",
       "      <td>1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S</td>\n",
       "      <td>3813</td>\n",
       "      <td>surface glycoprotein</td>\n",
       "      <td>21522</td>\n",
       "      <td>25334</td>\n",
       "      <td>1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYXXXXXXXXX...</td>\n",
       "      <td>3810</td>\n",
       "      <td>21452</td>\n",
       "      <td>25261</td>\n",
       "      <td>1</td>\n",
       "      <td>MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>828</td>\n",
       "      <td>ORF3a protein</td>\n",
       "      <td>25343</td>\n",
       "      <td>26170</td>\n",
       "      <td>1</td>\n",
       "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
       "      <td>828</td>\n",
       "      <td>25270</td>\n",
       "      <td>26097</td>\n",
       "      <td>1</td>\n",
       "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>E</td>\n",
       "      <td>228</td>\n",
       "      <td>envelope protein</td>\n",
       "      <td>26195</td>\n",
       "      <td>26422</td>\n",
       "      <td>1</td>\n",
       "      <td>MYSFVSEETGTLIVNSVLLFFAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
       "      <td>228</td>\n",
       "      <td>26122</td>\n",
       "      <td>26349</td>\n",
       "      <td>1</td>\n",
       "      <td>MYSFVSEEIGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>M</td>\n",
       "      <td>669</td>\n",
       "      <td>membrane glycoprotein</td>\n",
       "      <td>26473</td>\n",
       "      <td>27141</td>\n",
       "      <td>1</td>\n",
       "      <td>MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...</td>\n",
       "      <td>669</td>\n",
       "      <td>26400</td>\n",
       "      <td>27068</td>\n",
       "      <td>1</td>\n",
       "      <td>MAHSNGTITVEELKKLLEEWNLVIGFLFLAWICLLQFAYANRNRFL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>183</td>\n",
       "      <td>ORF6 protein</td>\n",
       "      <td>27152</td>\n",
       "      <td>27334</td>\n",
       "      <td>1</td>\n",
       "      <td>MHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTEN...</td>\n",
       "      <td>186</td>\n",
       "      <td>27079</td>\n",
       "      <td>27264</td>\n",
       "      <td>1</td>\n",
       "      <td>MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF7a protein</td>\n",
       "      <td>27341</td>\n",
       "      <td>27706</td>\n",
       "      <td>1</td>\n",
       "      <td>MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...</td>\n",
       "      <td>366</td>\n",
       "      <td>27271</td>\n",
       "      <td>27636</td>\n",
       "      <td>1</td>\n",
       "      <td>MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ORF7b</td>\n",
       "      <td>132</td>\n",
       "      <td>ORF7b</td>\n",
       "      <td>27703</td>\n",
       "      <td>27834</td>\n",
       "      <td>1</td>\n",
       "      <td>MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA</td>\n",
       "      <td>132</td>\n",
       "      <td>27633</td>\n",
       "      <td>27764</td>\n",
       "      <td>1</td>\n",
       "      <td>MIELSLIDFYLCFLAFLLLLVLIMLIIFWFSLELQDHNETCHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>366</td>\n",
       "      <td>ORF8 protein</td>\n",
       "      <td>27841</td>\n",
       "      <td>28206</td>\n",
       "      <td>1</td>\n",
       "      <td>MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "      <td>366</td>\n",
       "      <td>27771</td>\n",
       "      <td>28136</td>\n",
       "      <td>1</td>\n",
       "      <td>MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>N</td>\n",
       "      <td>1257</td>\n",
       "      <td>nucleocapsid phosphoprotein</td>\n",
       "      <td>28221</td>\n",
       "      <td>29477</td>\n",
       "      <td>1</td>\n",
       "      <td>MYNGPQNQRNGPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLPN...</td>\n",
       "      <td>1251</td>\n",
       "      <td>28151</td>\n",
       "      <td>29401</td>\n",
       "      <td>1</td>\n",
       "      <td>MSDNGPQNQRNALRITFGGPSDSTGSNQNGGARSKQRRPQGLPNNT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ORF10</td>\n",
       "      <td>117</td>\n",
       "      <td>ORF10 protein</td>\n",
       "      <td>29502</td>\n",
       "      <td>29618</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT</td>\n",
       "      <td>117</td>\n",
       "      <td>29426</td>\n",
       "      <td>29542</td>\n",
       "      <td>1</td>\n",
       "      <td>MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene  Length_x                      Product  Start_x  End_x  Strand_x  \\\n",
       "0   ORF1ab     13194           ORF1ab polyprotein      234  13427         1   \n",
       "1   ORF1ab     13194           ORF1ab polyprotein      234  13427         1   \n",
       "2   ORF1ab      8088           ORF1ab polyprotein    13427  21514         1   \n",
       "3   ORF1ab      8088           ORF1ab polyprotein    13427  21514         1   \n",
       "4   ORF1ab     13209            ORF1a polyprotein      234  13442         1   \n",
       "5        S      3813         surface glycoprotein    21522  25334         1   \n",
       "6    ORF3a       828                ORF3a protein    25343  26170         1   \n",
       "7        E       228             envelope protein    26195  26422         1   \n",
       "8        M       669        membrane glycoprotein    26473  27141         1   \n",
       "9     ORF6       183                 ORF6 protein    27152  27334         1   \n",
       "10   ORF7a       366                ORF7a protein    27341  27706         1   \n",
       "11   ORF7b       132                        ORF7b    27703  27834         1   \n",
       "12    ORF8       366                 ORF8 protein    27841  28206         1   \n",
       "13       N      1257  nucleocapsid phosphoprotein    28221  29477         1   \n",
       "14   ORF10       117                ORF10 protein    29502  29618         1   \n",
       "\n",
       "                                        Translation_x  Length_y  Start_y  \\\n",
       "0   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...     13194      164   \n",
       "1   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...      8088    13357   \n",
       "2   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...     13194      164   \n",
       "3   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...      8088    13357   \n",
       "4   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...     13209      164   \n",
       "5   MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYXXXXXXXXX...      3810    21452   \n",
       "6   MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...       828    25270   \n",
       "7   MYSFVSEETGTLIVNSVLLFFAFVVFLLVTLAILTALRLCAYCCNI...       228    26122   \n",
       "8   MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...       669    26400   \n",
       "9   MHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTEN...       186    27079   \n",
       "10  MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...       366    27271   \n",
       "11        MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA       132    27633   \n",
       "12  MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...       366    27771   \n",
       "13  MYNGPQNQRNGPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLPN...      1251    28151   \n",
       "14             MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT       117    29426   \n",
       "\n",
       "    End_y  Strand_y                                      Translation_y  \n",
       "0   13357         1  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "1   21444         1  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "2   13357         1  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "3   21444         1  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "4   13372         1  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "5   25261         1  MFVFLVLLPLVSSQCVMPLFNLITTTQSYTNSFTRGVYYPDKVFRS...  \n",
       "6   26097         1  MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...  \n",
       "7   26349         1  MYSFVSEEIGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...  \n",
       "8   27068         1  MAHSNGTITVEELKKLLEEWNLVIGFLFLAWICLLQFAYANRNRFL...  \n",
       "9   27264         1  MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...  \n",
       "10  27636         1  MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...  \n",
       "11  27764         1        MIELSLIDFYLCFLAFLLLLVLIMLIIFWFSLELQDHNETCHA  \n",
       "12  28136         1  MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...  \n",
       "13  29401         1  MSDNGPQNQRNALRITFGGPSDSTGSNQNGGARSKQRRPQGLPNNT...  \n",
       "14  29542         1             MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT  "
      ]
     },
     "execution_count": 429,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluating the common genes between the two genomes\n",
    "common_genes_df = pd.merge(older_genome_df, newer_genome_df, on=[\"Gene\", \"Product\"], how=\"inner\")\n",
    "common_genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ד.בחרו חמישה גנים משותפים וחשבו עבור כל גן את מדד ה Dnds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "target            0 MY--NGPQNQRNGP--RITFGGPSDSTGSNQNGERSGARSKQRRPQGLPNNTASWFTALT\n",
      "                  0 |---||||||||----|||||||||||||||||---||||||||||||||||||||||||\n",
      "query             0 M-SDNGPQNQRN--ALRITFGGPSDSTGSNQNG---GARSKQRRPQGLPNNTASWFTALT\n",
      "\n",
      "target           56 QHGKEDLKFPRGQGVPINTNSSPDDQIGYYRRATRRIRGGDGKMKDLSPRWYFYYLGTGP\n",
      "                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query            54 QHGKEDLKFPRGQGVPINTNSSPDDQIGYYRRATRRIRGGDGKMKDLSPRWYFYYLGTGP\n",
      "\n",
      "target          116 EAGLPYGANKDGIIWVATEGALNTPKDHIGTRNPANNAAIVLQLPQGTTLPKGFYAEGSR\n",
      "                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           114 EAGLPYGANKDGIIWVATEGALNTPKDHIGTRNPANNAAIVLQLPQGTTLPKGFYAEGSR\n",
      "\n",
      "target          176 GGSQASSRSSSRSRNSSRNSTPGSS-RGI-SPARMAGNGGDAALALLLLDRLNQLESKMS\n",
      "                180 |||||||||||||||||||||||||-|---|||||||||||||||||||||||.||||||\n",
      "query           174 GGSQASSRSSSRSRNSSRNSTPGSSKR--TSPARMAGNGGDAALALLLLDRLNKLESKMS\n",
      "\n",
      "target          234 GKGQQQQGQTVTKKSAAEASKKPRQKRTATKAYNVTQAFGRRGPEQTQGNFGDQELIRQG\n",
      "                240 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           232 GKGQQQQGQTVTKKSAAEASKKPRQKRTATKAYNVTQAFGRRGPEQTQGNFGDQELIRQG\n",
      "\n",
      "target          294 TDYKHWPQIAQFAPSASAFFGMSRIGMEVTPSGTWLTYTGAIKLDDKDPNFKDQVILLNK\n",
      "                300 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "query           292 TDYKHWPQIAQFAPSASAFFGMSRIGMEVTPSGTWLTYTGAIKLDDKDPNFKDQVILLNK\n",
      "\n",
      "target          354 HIDAYKTFPPTEPKKDKKKKADETQALPQRQKKQQTVTLLPAADLDDFSKQLQQSMSS-A\n",
      "                360 |||||||||||||||||||||||||||||||||||||||||||||||||||||||||--|\n",
      "query           352 HIDAYKTFPPTEPKKDKKKKADETQALPQRQKKQQTVTLLPAADLDDFSKQLQQSMS-RA\n",
      "\n",
      "target          413 DSTQA 418\n",
      "                420 ||||| 425\n",
      "query           411 DSTQA 416\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Align import substitution_matrices\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "seq1 = older_genome_df.loc[older_genome_df['Gene'] == 'N', 'Translation'].values[0]\n",
    "seq2 = newer_genome_df.loc[newer_genome_df['Gene'] == 'N', 'Translation'].values[0]\n",
    "\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "alignments = list(alignments)\n",
    "print(alignments[0])\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
